content/courses/ada-idioms/chapters/inheritance_idioms.rst

```diff
@@ -0,0 +1,577 @@
+legal, because clients would be promised capabilities that are not actually
+available.)
+
+When the parent type is tagged, the type derivation syntax for the child is a
+*type extension* declaration that introduces the child type's name, specifies
+the parent type, and then extends the parent representation with child-specific
+record components, if any. For example:
+
+.. code-block:: ada
+
+    type Child is new Parent with record ... end record;
+
+Even though the child type declaration does not include the reserved word
+:ada:`tagged` the child will be a tagged type because the parent type is tagged.
+The compiler would not allow the extension construct for a non-tagged parent
+type.
+
+Just as a private type can be visibly tagged or not, a private type can be
+*visibly derived* or not. When it is visibly derived, clients have a view of the
+private type that includes the fact of the derivation from the parent type.
+Otherwise, clients have no view of the parent type. Whether or not the child is
+visibly derived, the representation is not compile-time visible to clients, as
+for any private type. For example, type :ada:`Foo` is not visibly derived in the
+following:
+
+.. code-block:: ada
+
+    package P is
+       type Foo is tagged private;  --  visibly tagged but not visibly derived
+       -- ...
+    end P;
+
+To be visibly derived, we declare the child type as a private type using a
+*private extension*. A private extension is like a type extension, in that it
+introduces the child type name and the parent type. But like any private type
+declaration, it does not specify the type's representation. This is the first of
+the two steps for declaring a private type; hence it appears in the package
+visible part. For example:
+
+.. code-block:: ada
+
+    with ...
+    package P is
+       type Child is new Parent with private;  -- visibly derived from Parent
+    private
+       type Child is new Parent with record ... end record;
+    end P;
+
+The representation additions are not expressed until the private type's
+completion in the package private part, using a type extension. The steps are
+the same two for any private type: a declaration in the package visible part,
+with a completion in the package private part. The difference is the client
+visibility to the parent type.
+
+
+Solution
+--------
+
+There are two *solutions* in this entry, one for each of the two inheritance
+idioms under discussion. First, we will specify our building block choices,
+then we will show the two idiom expressions in separate subsections.
+
+- We will use tagged types for the sake of the full OOP capabilities. That is
+  the most common choice when inheritance is involved. The static form of
+  inheritance has cases in which it is useful, however those cases are very
+  narrow in applicability.
+
+- We will assume that both the parent type and the child type are both private
```

Comment:
"We will assume that both the parent" should be "We will assume that the parent"